runtime/interpreter/value.go

```diff
@@ -21200,6 +21200,10 @@ func NewAddressValueFromConstructor(
 }
 
 func ConvertAddress(memoryGauge common.MemoryGauge, value Value, locationRange LocationRange) AddressValue {
+	if address, ok := value.(AddressValue); ok {
```

Comment:
How can the value already be an address?
In the example below, converter function `Address(0x1)` takes an `Integer` parameter, so isn't the value always a subtype of Integer?

Reply:
I was also surprised about this at first. This isn't about the integer to address conversion in `Address(0x1)`, I removed that red herring in 91757526117cca9db042125cccb49687bc4249be. 

The case here is that in the return statement, the target type is `Address?` and the value type is `Address?`, and the value is an address value. The conversion function is still called because the optimization we currently have, only convert `if !valueType.Equal(unwrappedTargetType)` uses the _unwrapped_ type.

The cases for the conversion of other types, e.g. numeric types, already handle the case that the source and the target match.

This is unfortunately tech debt which we should clean up once we have 1.0 out of the door.

The next thing I found odd is that the value is actually an unboxed address – I was expecting that the conditional expression would already box the LHS, but that doesn't seem to be the case.
We might need to investigate that a bit more.

Reply:
ah okay, thanks for the explanation!